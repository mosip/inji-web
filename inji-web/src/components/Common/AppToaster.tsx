import React from "react";
import {ToastContainer} from "react-toastify";

export const AppToaster: React.FC = () => {
    return <ToastContainer
        position="top-right"
        autoClose={1000}
        hideProgressBar
        newestOnTop
        closeOnClick
        rtl={false}
        icon={<React.Fragment/>}
        pauseOnFocusLoss
        draggable
        pauseOnHover
        style={{width: '400px'}}
        theme="colored"
    />
}
