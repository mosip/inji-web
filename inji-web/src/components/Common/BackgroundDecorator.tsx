import React from 'react';

interface BackgroundDecoratorProps {
    circleCount?: number;
    topOffset?: number;
    baseRadius?: number;
    logoSrc?: string;
    logoAlt?: string;
    testId?: string;
}

export const BackgroundDecorator: React.FC<BackgroundDecoratorProps> = ({
    circleCount = 6,
    topOffset = 155,
    baseRadius = 96,
    logoSrc,
    logoAlt = "Logo",
    testId = "background-decorator"
}) => {
    return (
        <div 
            className="overflow-hidden absolute inset-0 z-0 flex items-center justify-center pointer-events-none"
            data-testid={testId}
        >
            <div className={`absolute top-[${topOffset}px]`} style={{ top: `${topOffset}px` }}>
                {[...Array(circleCount)].map((_, index) => {
                    const radius = baseRadius + index * baseRadius;
                    const opacity = 0.8 - index * 0.1;
                    return (
                        <div
                            key={index}
                            className="absolute rounded-full border overflow-hidden"
                            style={{
                                width: `${radius}px`,
                                height: `${radius}px`,
                                borderWidth: '1px',
                                borderColor: `rgba(228, 231, 236, ${opacity})`,
                                top: `calc(50% - ${radius / 2}px)`,
                                left: `calc(50% - ${radius / 2}px)`
                            }}
                            data-testid={`${testId}-circle-${index}`}
                        />
                    );
                })}
                
                {logoSrc && (
                    <div
                        className="flex items-center justify-center"
                        data-testid={`logo-${testId}-container`}
                    >
                        <img
                            src={logoSrc}
                            alt={logoAlt}
                            data-testid={`logo-${testId}`}
                        />
                    </div>
                )}
            </div>
        </div>
    );
};