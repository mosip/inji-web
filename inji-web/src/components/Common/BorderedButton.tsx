import React from "react";

export const BorderedButton:React.FC<BorderedButtonProps> = (props) => {
  return <div className={"bg-gradient-to-r rounded-lg from-iw-primary to-iw-secondary p-0.5"}>
            <div className="py-2 px-4 rounded-lg bg-white">
                <button
                    data-testid={props.testId}
                    onClick={props.onClick}
                    className="text-iw-tertiary font-bold rounded-lg">
                    {props.title}
                </button>
            </div>
    </div>
}

export type BorderedButtonProps = {
    testId: string;
    onClick: ()=> void;
    title: string;
}
