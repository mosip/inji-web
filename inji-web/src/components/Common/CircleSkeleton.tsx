type CircleSkeletonProps = {
    size?: string;
    className?: string;
};

export const CircleSkeleton = ({
                                   size = 'w-20 h-20',
                                   className = '',
                               }: CircleSkeletonProps) => {
    return (
        <div
            className={`rounded-full bg-gray-300 animate-pulse ${size} ${className}`}
        />
    );
};
  