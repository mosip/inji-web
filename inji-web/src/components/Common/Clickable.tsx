import React from 'react';

export interface ClickableProps {
    onClick: () => void;
    children: React.ReactNode;
    testId: string;
    className: string;
}

export const Clickable: React.FC<ClickableProps> = ({
                                                        onClick,
                                                        children,
                                                        testId,
                                                        className = '',
                                                    }) => {
    const handleKeyPress = (event: React.KeyboardEvent<HTMLDivElement>) => {
        if (event.key === 'Enter' || event.key === ' ') {
            event.preventDefault();
            onClick();
        }
    };

    return (
        <div
            role="menuitem"
            tabIndex={0}
            className={className}
            onClick={onClick}
            onKeyDown={handleKeyPress}
            data-testid={testId}
        >
            {children}
        </div>
    );
};
