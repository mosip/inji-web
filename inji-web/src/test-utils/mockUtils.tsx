import React, { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter, Route, BrowserRouter as Router, Routes } from 'react-router-dom';
import { reduxStore } from '../redux/reduxStore';

// Mock for storage module
export const mockStorageModule = () => {
    jest.mock('../utils/storage.ts', () => ({
        storage: {
            getItem: jest.fn(),
            setItem: jest.fn(),
            SESSION_INFO: 'SESSION_INFO',
            SELECTED_LANGUAGE: 'selectedLanguage'
        }
    }));
};

// Mock for localStorage
export const mockLocalStorage = () => {
    let store: { [key: string]: string } = {};
  
    const localStorageMock = {
      getItem: jest.fn((key: string) => store[key] || null),
      setItem: jest.fn((key: string, value: string) => {
        store[key] = value;
      })
    };
  
    Object.defineProperty(window, 'localStorage', { 
      value: localStorageMock,
      writable: true
    });
  
    return localStorageMock;
  };

// API Mock
export const mockApi = {
    authorizationRedirectionUrl: 'http://mockurl.com'
};

// Crypto Mock
export const mockCrypto = {
    getRandomValues: (array: Uint32Array) => {
        for (let i = 0; i < array.length; i++) {
            array[i] = 4294967296;
        }
        return array;
    },
    subtle: {} as SubtleCrypto,
    randomUUID: () => '123e4567-e89b-12d3-a456-426614174000'
} as Crypto;

export const mockUseTranslation = () => {
    jest.mock('react-i18next', () => ({
        useTranslation: () => ({
            t: (key: string) => key,
        }),
    }));
};

export const mockUseNavigate = () => {
    const mockNavigate = jest.fn();
    jest.mock('react-router-dom', () => ({
        ...jest.requireActual('react-router-dom'),
        useNavigate: mockNavigate,
    }));
};

export const mockUseSelector = () => {
    jest.mock('react-redux', () => ({
        ...jest.requireActual('react-redux'),
        useSelector: jest.fn(),
    }));
};

export const setMockUseSelectorState = (state: any) => {
    const useSelectorMock = require('react-redux').useSelector;
    useSelectorMock.mockImplementation((selector: any) => selector(state));
};

export const mockUseGetObjectForCurrentLanguage = () => {
    jest.mock('../utils/i18n', () => ({
        getObjectForCurrentLanguage: jest.fn(),
    }));
};

export const wrapUnderRouter = (children: React.ReactNode) => {
    return <Router>{children}</Router>;
};

export const mockUseSearchCredentials = () => {
    jest.mock('../components/Credentials/SearchCredential', () => ({
        SearchCredential: () => <></>,
    }));
 };

export const mockUseDispatch = () => {
    jest.mock('react-redux', () => ({
        ...jest.requireActual('react-redux'),
        useDispatch: jest.fn(),
    }));
};
              
export const mockUseSpinningLoader = () => {
    jest.mock('../components/Common/SpinningLoader', () => ({
        SpinningLoader: () => <></>,
    }));
};

export const mockUseLanguageSelector = () => {
    jest.mock('../components/Common/LanguageSelector', () => ({
        LanguageSelector: () => <></>,
    }));
};

interface RenderWithProviderOptions extends Omit<RenderOptions, 'queries'> {}

export const renderWithProvider = (element: ReactElement, options?: RenderWithProviderOptions) => {
    return render(
        <Provider store={reduxStore}>
            <Router>
                {element}
            </Router>
        </Provider>,
        options
    );
};
    
export const mockUseParams = ()=>{
    jest.mock('react-router-dom', () => ({
        ...jest.requireActual('react-router-dom'),
        useParams: () => ({ issuerId: 'test-issuer-id' }),
      }));   
};

export const mockUseapiObject = () =>{
    jest.mock('../utils/api', () => ({
        api: {
          mimotoHost: 'http://mocked-api-host',
        },
      }));
};

export const mockUseFetch = () =>{
    jest.mock('../hooks/useFetch', () => {
        const RequestStatus = {
          LOADING: 'LOADING',
          DONE: 'DONE',
          ERROR: 'ERROR',
        };
        return {
          useFetch: () => jest.fn(),
          RequestStatus,
        };
      });
};

export const mockUseToast = () =>{
    jest.mock('react-toastify', () => ({
      toast: {
        error: jest.fn(),
      },
    }));
};

export const mockusemisc = ()=>{
    jest.mock('../utils/misc', () => ({
        downloadCredentialPDF: jest.fn(),
        getErrorObject: jest.fn(),
        getTokenRequestBody: jest.fn(),
      }));
};

export const mockusei18n = ()=>{
    jest.mock('react-i18next', () => ({
        useTranslation: () => ({
          t: (key: string) => key,
        }),
        initReactI18next: {
          type: '3rdParty',
          init: jest.fn(),
        },
      }));

};

export const mockWindowLocation = (url: string) => {
    const location = new URL(url) as unknown as Location;
    Object.defineProperty(window, 'location', {
      value: location,
      writable: true,
    });
  };

export const renderWithRouter = (Element: React.ReactElement, { route = '/' } = {}) => {
    window.history.pushState({}, 'Test page', route);
    return render(
      <BrowserRouter>
        <Provider store={reduxStore}>
          <Routes>
            <Route path="*" element={Element} />
          </Routes>
        </Provider>
      </BrowserRouter>
    );
  };

