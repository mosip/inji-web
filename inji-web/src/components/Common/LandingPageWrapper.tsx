import React from "react";
import {useNavigate} from "react-router-dom";
import {useTranslation} from "react-i18next";
import {BorderedButton} from "./Buttons/BorderedButton";
import {RouteValue} from "../../types/data";
import {ROUTES} from "../../utils/constants";
import {useUser} from "../../hooks/useUser";

interface LandingPageWrapperProps {
    icon: React.ReactNode;
    title: string;
    subTitle?: string;
    gotoHome: boolean;
    navUrl?: RouteValue;
};

export const LandingPageWrapper: React.FC<LandingPageWrapperProps> = (props) => {
    const navigate = useNavigate();
    const {t} = useTranslation("RedirectionPage");
    const {isUserLoggedIn} = useUser();
    const wrapperConfig = isUserLoggedIn
        ? {
            navUrl: ROUTES.USER_HOME,
            classNames: {
                outerContainer: "flex flex-col justify-center items-center",
                titleContainer: "mt-5",
                subTitleContainer: "my-3 px-8 text-center",
            },
        }
        : {
            navUrl: ROUTES.ROOT,
            classNames: {
                outerContainer: "flex flex-col justify-center items-center pt-32",
                titleContainer: "my-4",
                subTitleContainer: "mb-6 px-10 text-center",
            },
        };

    const {navUrl = wrapperConfig.navUrl} = props;

    return (
        <div data-testid="outer-container-download-result" className={wrapperConfig.classNames.outerContainer}>
            {props.icon}
            <div className={wrapperConfig.classNames.titleContainer} data-testid="title-container-download-result">
                <p className="font-bold" data-testid="title-download-result">
                    {props.title}
                </p>
            </div>
            {props.subTitle &&
                <div className={wrapperConfig.classNames.subTitleContainer}
                     data-testid="subtitle-container-download-result">
                    <p data-testid="subtitle-download-result">{props.subTitle}</p>
                </div>
            }
            {props.gotoHome && (
                <BorderedButton
                    testId="btn-home-download-result"
                    onClick={() => navigate(navUrl)}
                    title={t("navigateButton")}
                />
            )}
        </div>
    );
};