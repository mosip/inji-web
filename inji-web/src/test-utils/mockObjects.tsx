import { IssuerConfigurationObject } from '../types/data';
import { LandingPageWrapperProps } from '../components/Common/LandingPageWrapper';
import { NavBarProps } from '../types/components';
export const mockIssuerDisplayArrayObject = {
    name: "Name",
    language: "en",
    locale: "en",
    logo: {
        url: "https://url.com",
        alt_text: "alt text of the url"
    },
    title: "Title",
    description: "Description",
};

export const mockCredentialTypeDisplayArrayObject = {
    name: "Name",
    locale: "en",
    logo: "https://url.com",
};

export const mockCredentials: IssuerConfigurationObject = {
    credentials_supported: [
        {
            name: "InsuranceCredential",
            scope: "mosip_vc_ldp",
            display: [
                {
                    name: "Health Insurance",
                    locale: "en",
                    logo: "https://url.com"
                }
            ]
        },
        {
            name: "AnotherCredential",
            scope: "mosip_ldp_vc",
            display: [
                {
                    name: "Another Credential",
                    locale: "en",
                    logo: "https://url.com"
                }
            ]
        }
    ],
    "authorization_endpoint": "https://env.net/authorize",
    "grant_types_supported": ["authorization_code"]
};


export const mockIssuerObject = {
        name: 'Test Issuer',
        desc: 'Test Description',
        protocol: 'OTP' as 'OTP', // Explicitly set the type to 'OTP' or 'OpenId4VCI'
        issuer_id: 'test-issuer',
        authorization_endpoint: 'https://auth.test.com',
        credentials_endpoint: 'https://credentials.test.com',
        display: [{
            name: "Name",
            language: "en",
            locale: "en",
            logo: {
                url: "https://url.com",
                alt_text: "alt text of the url"
            },
            title: "Title",
            description: "Description",
        }],
        client_id: 'test-client-id',
        redirect_uri:'test-redirect-uri',
        token_endpoint:'test-token_endpoint',
        proxy_token_endpoint:'test-proxy_token_endpoint',
        client_alias:'',
        ovp_qr_enabled: true,
        scopes_supported: ['scope1', 'scope2']
}

export const mockIssuerObjectList =  [
    {
        name: 'Issuer 1',
        desc: 'Description 1',
        protocol: 'OpenId4VCI',
        issuer_id: 'issuer1',
        authorization_endpoint: 'https://issuer1.com/auth',
        credentials_endpoint: 'https://issuer1.com/credentials',
        display: [
            {
                name: 'Issuer 1',
                language: 'en',
                locale: 'en-US',
                logo: { url: 'https://issuer1.com/logo.png', alt_text: 'Issuer 1 Logo' },
                title: 'Issuer 1 Title',
                description: 'Issuer 1 Description',
            },
        ],
        client_id: 'client1',
        redirect_uri: 'https://issuer1.com/redirect',
        token_endpoint: 'https://issuer1.com/token',
        proxy_token_endpoint: 'https://issuer1.com/proxy-token',
        client_alias: 'client1-alias',
        ovp_qr_enabled: true,
        scopes_supported: ['openid', 'profile'],
    },
    {
        name: 'Issuer 2',
        desc: 'Description 2',
        protocol: 'OpenId4VCI',
        issuer_id: 'issuer2',
        authorization_endpoint: 'https://issuer2.com/auth',
        credentials_endpoint: 'https://issuer2.com/credentials',
        display: [
            {
                name: 'Issuer 2',
                language: 'en',
                locale: 'en-US',
                logo: { url: 'https://issuer2.com/logo.png', alt_text: 'Issuer 2 Logo' },
                title: 'Issuer 2 Title',
                description: 'Issuer 2 Description',
            },
        ],
        client_id: 'client2',
        redirect_uri: 'https://issuer2.com/redirect',
        token_endpoint: 'https://issuer2.com/token',
        proxy_token_endpoint: 'https://issuer2.com/proxy-token',
        client_alias: 'client2-alias',
        ovp_qr_enabled: false,
        scopes_supported: ['openid', 'profile'],
    },
]

export const mockLandingPageWrapperProps: LandingPageWrapperProps = {
    icon: <div data-testid="Test-Icon">Icon</div>,
    title: "Test Title",
    subTitle: "Test SubTitle",
    gotoHome: true,
};

export const mockNavBarProps: NavBarProps = {
    title: 'Test Title',
    link: '/test-link',
    search: true,
};