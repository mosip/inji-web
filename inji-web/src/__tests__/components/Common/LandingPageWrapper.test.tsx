import React from 'react';
import {  screen, fireEvent } from '@testing-library/react';
import { LandingPageWrapper } from '../../../components/Common/LandingPageWrapper';
import {  mockUseTranslation, renderWithProvider,mockUseNavigate} from '../../../test-utils/mockUtils';
import { mockLandingPageWrapperProps } from '../../../test-utils/mockObjects';


// Mock useTranslation
mockUseTranslation();

//todo : extract the local method to mockUtils, which is added to bypass the routing problems
const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
}));

describe("Testing the Layout of LandingPageWrapper", () => {
    test('Check if the layout is matching with the snapshots', () => {
        const { asFragment } = renderWithProvider(<LandingPageWrapper {...mockLandingPageWrapperProps} />);
        expect(asFragment()).toMatchSnapshot();
    });
});

describe("Testing the Functionality LandingPageWrapper", () => {
    beforeEach(()=>{
        mockUseNavigate();
    })
    test('Check it navigates to home when the home button is clicked', () => {
        renderWithProvider(<LandingPageWrapper {...mockLandingPageWrapperProps} />);
        const homeButton = screen.getByTestId("btn-home-download-result");
        fireEvent.click(homeButton);
        expect(mockNavigate).toHaveBeenCalledWith('/'); 
    });

    test('Check if it have the Title and the SubTitle', () => {
        renderWithProvider(<LandingPageWrapper {...mockLandingPageWrapperProps} />);
        expect(screen.getByTestId("title-download-result")).toHaveTextContent("Test Title");
        expect(screen.getByTestId("subtitle-container-download-result")).toHaveTextContent("Test SubTitle");
    });
    afterEach(()=>{
        jest.clearAllMocks();
    })
});
