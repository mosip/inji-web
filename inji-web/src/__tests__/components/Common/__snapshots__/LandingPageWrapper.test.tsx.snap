// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the Layout of LandingPageWrapper Check if the layout is matching with the snapshots 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col justify-center items-center pt-32"
    data-testid="outer-container-download-result"
  >
    <div
      data-testid="Test-Icon"
    >
      Icon
    </div>
    <div
      class="my-4"
      data-testid="title-container-download-result"
    >
      <p
        class="font-bold"
        data-testid="title-download-result"
      >
        Test Title
      </p>
    </div>
    <div
      class="mb-6 px-10 text-center"
      data-testid="subtitle-container-download-result"
    >
      <p
        data-testid="subtitle-download-result"
      >
        Test SubTitle
      </p>
    </div>
    <div
      class=""
    >
      <div
        class="bg-gradient-to-r rounded-lg text-center cursor-pointer shadow hover:shadow-lg from-iw-primary to-iw-secondary p-0.5"
      >
        <div
          class="py-2 px-4 rounded-lg bg-white"
        >
          <button
            class="text-iw-tertiary font-bold rounded-lg"
            data-testid="btn-home-download-result"
          >
            <span
              class="bg-gradient-to-r from-iw-primary to-iw-secondary bg-clip-text text-md font-bold text-transparent"
            >
              Go To Home
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
